pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -21,50 +36,491 @@ func init() {
 //
 // [Rego]: https://www.openpolicyagent.org/docs/latest/policy-language/
-type RegoEnforcer struct{}
+type RegoEnforcer struct {
+	// Rego which describes policy behavior (see above)
+	behavior string
+	// Rego which describes policy objects (containers, etc.)
+	objects string
+	// Mutex to prevent concurrent access to fields
+	mutex *sync.Mutex
+	// Rego data object, used to store policy state
+	data map[string]interface{}
+	// Base64 encoded (JSON) policy
+	base64policy string
+	// Compiled modules
+	compiledModules *ast.Compiler
+}
 
-var (
-	_                 SecurityPolicyEnforcer = (*RegoEnforcer)(nil)
-	ErrNotImplemented                        = errors.New("not implemented")
-)
+var _ SecurityPolicyEnforcer = (*RegoEnforcer)(nil)
 
-func createRegoEnforcer(_ SecurityPolicyState, _, _ []oci.Mount) (SecurityPolicyEnforcer, error) {
-	return &RegoEnforcer{}, nil
+type securityPolicyInternal struct {
+	AllowAll   bool
+	Containers []*securityPolicyContainer
 }
 
-func (RegoEnforcer) EnforceDeviceMountPolicy(_, _ string) error {
-	return ErrNotImplemented
+func (sp SecurityPolicy) toInternal() (*securityPolicyInternal, error) {
+	policy := new(securityPolicyInternal)
+	var err error
+	if policy.Containers, err = sp.Containers.toInternal(); err != nil {
+		return nil, err
+	}
+
+	return policy, nil
 }
 
-func (RegoEnforcer) EnforceDeviceUnmountPolicy(_ string) error {
-	return ErrNotImplemented
+func toOptions(values []string) Options {
+	elements := make(map[string]string)
+	for i, value := range values {
+		elements[fmt.Sprint(i)] = value
+	}
+	return Options{
+		Length:   len(values),
+		Elements: elements,
+	}
 }
 
-func (RegoEnforcer) EnforceOverlayMountPolicy(_ string, _ []string) error {
-	return ErrNotImplemented
+func (mounts *Mounts) Append(other []oci.Mount) {
+	start := mounts.Length
+	for i, mount := range other {
+		mounts.Elements[fmt.Sprint(i+start)] = Mount{
+			Source:      mount.Source,
+			Destination: mount.Destination,
+			Type:        mount.Type,
+			Options:     toOptions(mount.Options),
+		}
+	}
+
+	mounts.Length += len(other)
 }
 
-func (RegoEnforcer) EnforceCreateContainerPolicy(_ string, _, _ []string, _ string) error {
-	return ErrNotImplemented
+type StringArray []string
+
+func (array StringArray) MarshalRego() string {
+	values := make([]string, len(array))
+	for i, value := range array {
+		values[i] = fmt.Sprintf("\"%s\"", value)
+	}
+
+	return fmt.Sprintf("[%s]", strings.Join(values, ","))
+}
+
+func writeCommand(builder *strings.Builder, command []string, indent string) error {
+	array := (StringArray(command)).MarshalRego()
+	_, err := builder.WriteString(fmt.Sprintf("%s\"command\": %s,\n", indent, array))
+	return err
 }
 
-func (RegoEnforcer) EnforceWaitMountPointsPolicy(_ string, _ *oci.Spec) error {
-	return ErrNotImplemented
+func (e EnvRuleConfig) MarshalRego() string {
+	return fmt.Sprintf("{\"pattern\": \"%s\", \"strategy\": \"%s\"}", e.Rule, e.Strategy)
```

Comment:
Go has raw strings (enclosed in backticks) so you don't need to escape the quotes. 
```suggestion
	return fmt.Sprintf(`{"pattern": "%s", "strategy": "%s"}"`, e.Rule, e.Strategy)
```

Reply:
This appears to have broken things, not sure why. I've reverted for now. Perhaps @matajoh might have an idea. It's easy to reproduce by making the change. Running the marshall rego test exhibits the failure.

Reply:
Dur. I see the bug. There's an extra " in your suggestion.